C++/C.sublime-syntax

```diff
@@ -13,1427 +14,893 @@ first_line_match: |-
+      captures:
+        1: entity.name.function.c
+        2: meta.group.c punctuation.section.group.begin.c
+      push:
+        - function-tail
+        - function-params
+    - match: \b{{identifier}}\b
+  function-params:
+    - meta_content_scope: meta.function.parameters.c meta.group.c
+    - match: \)
+      scope: meta.group.c punctuation.section.group.end.c
+      pop: 1
+    - match: \}|\] # Fallback
+      scope: invalid.illegal.stray-bracket-end.c
+      pop: 2
+    - match: ; # Fallback
+      scope: invalid.illegal.unexpected-terminator.c
+      pop: 2
+
+    - match: (?={{identifier}}\s*(?!\s*[\[,\)]|$))
+      push:
+        - match: \b
+          set: function-params-name
+        - include: expression
+    - include: expression
+  function-params-name:
+    - match: (?=[\]\}\);]) # Fallback
+      pop: 1
+    - match: '{{fn_pointer}}'
       captures:
-        1: punctuation.definition.comment.c
+        1: keyword.operator.c
+        2: variable.parameter.c
+        3: punctuation.section.group.begin.c
+      set: function-params
+    - include: expression-identifiers
+    - match: ({{identifier}})\s*(\[)
+      captures:
+        1: variable.parameter.c
+        2: meta.brackets.square.c punctuation.section.brackets.begin.c
+      set: enclosure.brackets
+    - match: ({{identifier}})\s*(?=[,\)]|$)
+      captures:
+        1: variable.parameter.c
+      pop: 1
+    - include: expression
+  function-tail:
+    - meta_scope: meta.function.c
+    - match: (?=;)
+      pop: 1
+    - match: \{
+      scope: punctuation.section.block.begin.c
+      set: function-body
+    - include: expression-identifiers
+    - match: \[\[
+      scope: meta.brackets.square.c punctuation.section.brackts.begin.c
+      push: enclosure.attribute
+  function-body:
+    - meta_scope: meta.function.c meta.block.c
+    - match: \}
+      scope: punctuation.section.block.end.c
+      pop: 1
+    - include: expression
+    - include: label
 
-  block-comment-body:
-    - meta_include_prototype: false
-    - meta_scope: comment.block.c
-    - match: \*/
-      scope: punctuation.definition.comment.end.c
-      pop: true
-    - match: ^\s*(\*)(?!\**/)
+  expression-identifiers:
+    - include: operator.word
+    - include: type
+    - include: keyword
+    - include: constant
+
+###[ SIMPLE STRUCTURES ]############################################################
+
+  enum-def: # for C++ to override
+    - match: \benum\b
+      scope: keyword.declaration.enum.c
+      branch_point: enum
+      branch: [type.enum, pop-immediately]
+
+  type:
+    - match: \btypedef\b
+      scope: keyword.declaration.type.c
+      push: type.typedef
+    - include: enum-def
+    - match: \bstruct\b
+      scope: keyword.declaration.struct.c
+      branch_point: struct
+      branch: [type.struct, pop-immediately]
+    - match: \bunion\b
+      scope: keyword.declaration.union.c
+      branch_point: union
+      branch: [type.union, pop-immediately]
```

Comment:
Maybe organize all sorts of definitions in separate contexts, grouped logically? This way we don't need to introduce exceptions like `enum-def` in case an extending syntax needs to override it.